extent smooth rendering to mesh layer

vcloarec opened the pull request:
As for raster layers (https://github.com/qgis/QGIS/pull/40960) and vector layers (https://github.com/qgis/QGIS/pull/41198), this PR allows the "smooth" rendering aka "smart map redraw" working also for mesh layer. 
Quite straight forward after the work of @PeterPetrik  and @nyalldawson.
This smooth navigation is also extended for the temporal dimension.

PeterPetrik commented on src/core/mesh/qgsmeshlayertemporalproperties.cpp:
```diff
@@ -81,6 +81,11 @@ QgsDateTimeRange QgsMeshLayerTemporalProperties::calculateTemporalExtent( QgsMap
 
+QgsTemporalProperty::Flags QgsMeshLayerTemporalProperties::flags() const
+{
+  return QgsTemporalProperty::Flags() | QgsTemporalProperty::FlagDontInvalidateCachedRendersWhenRangeChanges;
```
FlagDontInvalidateCachedRendersWhenRangeChanges -> FlagDontInvalidateRendererCacheWhenRangeChanges

nyalldawson replied:
This isn't the correct use of this flag -- this flag was designed with a single use case in mind, where an entire layer's visibility is controlled by the temporal framework but where the actual appearance of that layer doesn't change regardless of what the actual range is. I.e. it's used only for rasters and vectors set to display for a fixed time range only -- but avoid the cache invalidation we ensure that these layers don't need to be needlessly re-rendered when the time range changes, but instead just selectively use the cached image or not based on the layer visibility...

PeterPetrik commented on src/core/qgsmaprenderercache.h:
```diff
@@ -159,6 +159,7 @@ class CORE_EXPORT QgsMapRendererCache : public QObject
       QgsWeakMapLayerPointerList dependentLayers;
       QgsRectangle cachedExtent;
       QgsMapToPixel cachedMtp;
+      QgsDateTimeRange cacheTimeRange;
```
cacheTimeRange > cachedTimeRange

PeterPetrik commented on tests/src/core/testqgsmaprenderercache.cpp:
```diff
@@ -100,7 +100,7 @@ void TestQgsMapRendereCache::testCache()
   QgsRectangle extent2( 20, 0, 120, 200 );
   QgsMapToPixel mtp2( 1, 70, 50, 100, 100, 0.0 ); // move by 20
-  cache.updateParameters( extent2, mtp2 );
+  cache.updateParameters( extent2, mtp2, QgsDateTimeRange() );
```
Please add a test with the changed time range too please

PeterPetrik commented on src/core/qgsmaprenderercache.cpp:
```diff
@@ -105,7 +105,7 @@ bool QgsMapRendererCache::init( const QgsRectangle &extent, double scale )
 }
 
-bool QgsMapRendererCache::updateParameters( const QgsRectangle &extent, const QgsMapToPixel &mtp )
+bool QgsMapRendererCache::updateParameters( const QgsRectangle &extent, const QgsMapToPixel &mtp, const QgsDateTimeRange &timeRange )
```
you need to reset mTimeRange in  QgsMapRendererCache::init too (in deprecated function)

or @nyalldawson do you think we can remove the deprecated function at all ? I am starting to have doubts if it works at all now and QgsMapRendererCache is not something you use from plugins..

PeterPetrik reviewed (commented):
Thanks! Can you add screencast too please?

nyalldawson commented:
@vcloarec can you explain the rationale behind the temporal handling in the cache? AFAIKS it all should be handled outside of the cache by the logic in map canvas...

vcloarec commented:
@nyalldawson 

Indeed, updating the map is handled in the map canvas by cleaning the cache before new rendering, but this cleaning avoids to use the previously rendered image when waiting for the renderer job, as for non-temporal navigation (pan or zoom).
That leads to flickering with mesh layer with temporal navigation (and it seems for vector layer too, but I do not try a lot...).

As this cleaning is not done when there is the flag QgsTemporalProperty::FlagDontInvalidateCachedRendersWhenRangeChanges, I have implemented temporal properties of mesh layers to return always this flag. Indeed, for mesh layer, maybe I missed something, but I don't see any reason for cleaning the cache except to refresh the rendering when time range is changed (but that leads to flickering, so finally not wanted...). Then, to allow refreshing of the final rendered image when time range changes, it is necessary to introduce the time range in the cache, as for map extent and the mapToPixel. After all, it is also a parameter that defines the rendering...
But, indeed, harmful consequences is that the cache will never be used anymore to avoid rendering after temporal range changes even for static layer... not very good...

So, for now, I see two solutions :
- give up, for now, this smooth rendering for temporal navigation (pity because this flickering is quite annoying, it appears when the rendering is greater than the map update interval that is 250 ms by default)
OR
- add a flag in the CacheParameters to check if the layer is temporal or not. If not, the cache does not care about the temporal range, if yes, cache image is not considered if temporal range is different (implemented in the last commit)

Any suggestion is welcome!

vcloarec commented:
screen cast
### **pan navigation**
**before:**

![fleekering_before](https://user-images.githubusercontent.com/7416892/106183213-54b33d00-6176-11eb-89e2-e295b4b77908.gif)

**After:**

![fleekering_after](https://user-images.githubusercontent.com/7416892/106183262-63015900-6176-11eb-88e6-68ddf3b24c1f.gif)


### Temporal navigation

**before:**

![fleekering_temporal_before](https://user-images.githubusercontent.com/7416892/106183338-7ad8dd00-6176-11eb-9833-28aeb5bc840c.gif)

**After:**

![fleekering_temporal_after](https://user-images.githubusercontent.com/7416892/106183375-85937200-6176-11eb-82ae-fea768f67202.gif)

nyalldawson commented:
Sorry - I didn't get a chance to revisit this today... I'll review Monday!

PeterPetrik commented:
@vcloarec can you rebase please?

PeterPetrik commented on tests/src/core/testqgsmaprenderercache.cpp:
```diff
@@ -97,10 +101,20 @@ void TestQgsMapRendereCache::testCache()
     QVERIFY( img.pixelColor( 10, 20 ) == Qt::blue );
   }
 
+  // same extent and different time ange
```
typo "ange"

vcloarec commented:
rebase done
 (manual one considering the big conflict with https://github.com/qgis/QGIS/pull/41220)

vcloarec commented:
I am wondering if this PR, with this one (https://github.com/qgis/QGIS/pull/41220), does not make useless the clear of the cache with QgsMapCanvas::clearTemporalCache() (introduced by https://github.com/qgis/QGIS/pull/36277) and the stuff with QgsTemporalProperty::FlagDontInvalidateCachedRendersWhenRangeChanges.
@nirvn ?

vcloarec commented:
Travis failing is unrelated, any blocking point? @nyalldawson

nyalldawson commented:
Tab is open to review -- please don't merge yet!

nyalldawson commented on src/core/mesh/qgsmeshlayerrenderer.cpp:
```diff
@@ -300,6 +303,7 @@ bool QgsMeshLayerRenderer::render()
   renderMesh();
   renderVectorDataset();
 
+  mReadyToCompose = true;
```
Are mesh layers ever slow to redraw? I'm wondering if a similar approach as is done for vector layers would be beneficial here, where we set a timer when the rendering starts and if it isn't complete in 3 seconds then we set mReadyToComposer to true early so that the progressive rendering updates are shown

vcloarec replied:
I think it could be interested only for the mesh frame or arrows of the vectors datasets (not for streamlines or traces that are actually rendered a bit like raster, that is, by producing an image). But as these renderings can be disabled or set to be coarser, I don't know if it  is good to add more complexity only for these renderings.
I will look how it is complex to add this for arrow and mesh frame...

nyalldawson reviewed (changes requested):
I still am unsure about the need for the cache to know anything about the time range -- I suspect you've only had to do that here because you've used the FlagDontInvalidateCachedRendersWhenRangeChanges incorrectly, and so have bypassed all the existing temporal cache interactions...

vcloarec commented:
@nyalldawson  I think I was not clear, sorry for that. I will try to be clearer, sorry for long text.

The issue I try to fix here is not only for mesh layer but for others too. The problem is when the rendering of the layer is longer than the the "map update interval" the user can change in "Option"->"Settings"->tab "Rendendering", this value is "only" 250 ms by default. Even the rendering is just longer than this value, when this "map update interval" is out during rendering layers, the rendering of the map canvas use the cache to render the map:
https://github.com/qgis/QGIS/blob/f9db58a5d1843c272864e283f4a6a9bbc3b50f88/src/gui/qgsmapcanvas.cpp#L1003
But this cache was cleared before the rendering job start because the temporal range has changed: https://github.com/qgis/QGIS/blob/f9db58a5d1843c272864e283f4a6a9bbc3b50f88/src/gui/qgsmapcanvas.cpp#L973 
so,  when layer rendering is longer than "map update interval", blank image is rendered and we have this flickering. And this even for a vector layer. Here a screencast with a temporal vector layer rendered with master and a rendering longer than 300 ms ("map update interval" = 250 ms).


![vector_layer_flickering](https://user-images.githubusercontent.com/7416892/106786226-b109d700-6624-11eb-9bac-852828e15951.gif)


To share my experience, I don't know for vector layer, but when working with a good machine with Linux, rendering of mesh layer is very often under this value, so I've never really seen the flickering ... except recently when I demonstrate mesh layer for teaching a client about hydraulic modelling, and despite his computer was not so bad, but under Windows, temporal visualization was quite bad... I was a bit ashamed about that poor rendering, and I feel myself very uncomfortable when I imagine this rendering for almost all the users... Not sure this flickering was here before the new temporal framework. Temporal dimension is essential for mesh layer, I think more than for other layer types. So I investigated and I thought that was only coming from the slow Windows rendering and I fixed that (https://github.com/qgis/QGIS/pull/40821). But for this PR, I see that it also comes from this cache cleaning, and with not so good machine, this flickering will happen...

So, how to fix this issue? (I think a quite big problem for mesh layer as temporal dimension is so important)
With the actual implementation of the cache, it is necessary to clear the cache for temporal layers when changing temporal range, so the rendering of layers can be refresh even the extent does not change.
To avoid this cleaning for mesh layer, with opportunity, I used the existent flag 'FlagDontInvalidateCachedRendersWhenRangeChanges' that allows, as this name says: "Don't Invalidate Cache Renders When Range Change", after all, precisely what I needed! Seeing the code around this flag, indeed, I thought that was not exactly why this flag was introduced, but that allows to fix mesh rendering without change the logic for other layer types for which I was not sure about the logic.
Then, to be sure the mesh layer rendering is updated in the cache, I introduced this temporal condition in the cache.

> I still am unsure about the need for the cache to know anything about the time range 

Why does cache not have to know about the time range? Map extent is the spatial characteristic and map time range is the temporal one. 
How to avoid the cache for temporal layers is invalidate while we need it?
How to clear it, or better, replace it with a more adapted one when it is really necessary? https://github.com/qgis/QGIS/pull/41220 help a lot for this if we add the time range criteria in the cache
Where is the better place to handle the interaction time range / cache if not in the cache itself? Personally, I find simpler and proper to do that encapsulate in the cache
Don't you think that handling the time range in the cache for temporal layer (especially with https://github.com/qgis/QGIS/pull/41220) will simplify the cache handling in the map canvas? No need anymore to clear the cache only for temporal layer from map canvas, and we can use the flag 'FlagDontInvalidateCachedRendersWhenRangeChanges' in the cache (may be replaced by "DontUpdateCacheRendererWhenRangeChanges") only for what it was introduced.

Also, if the time range is only handled in the cache, the flickering for vector layer can also be fixed. If I disable the cache clearing from map canvas in my branch the result for my temporal vector layer is (same condition):

![vector_layer_flickering_after](https://user-images.githubusercontent.com/7416892/106786255-b9621200-6624-11eb-91e8-53eafff80222.gif)

I am open to another solution to handle this temporal cache stuff, but I find this one quite simple and efficient.

vcloarec commented:
@nyalldawson 
Any chance we can advance on this for 3.18? 
Should we give up for no for this smoother temporal navigation? Rework on it for all layers type with time range criteria in the cache?

nyalldawson commented:
@vcloarec 
> Any chance we can advance on this for 3.18?

I'd like to, but I'm concerned that we're getting too close to release to be doing major changes to these classes now. Could you split off the mesh redraw improvement alone and we can leave the cache temporal handling rework till after release? 

> To share my experience, I don't know for vector layer, but when working with a good machine with Linux, rendering of mesh layer is very often under this value, so I've never really seen the flickering ... except recently when I demonstrate mesh layer for teaching a client about hydraulic modelling, and despite his computer was not so bad, but under Windows, temporal visualization was quite bad... 

Totally valid point, and definitely something it would be nice to improve. But just for reference we should keep in mind that it's not a regression and that if we release 3.18 like this it won't be any step backward from 3.16 or other earlier releases.

>Should we give up for no for this smoother temporal navigation? Rework on it for all layers type with time range criteria in the cache?

No, we definitely shouldn't give up on it, but I do think a deeper rethink is necessary. My gut feeling now is that we should extend the layer cache to allow for multiple images per layer, and use a maximum cache size approach to invalidate and discard the image which hasn't been used for the longest time when we need to. This would give us two benefits:

1. For temporal layers, we could then follow your approach and add the time range to the cached images. We'd then be able to store more than one rendered image for different time ranges in the cache -- which would give a huge benefit to rendering speed of animations (especially those using WMST services, and complex vector layers), because we wouldn't be redrawing frames which we've previously rendered, we'd just be grabbing them from the cache. 
2. We could store multiple images with different extents for a single layer -- this means that the current "out of frame background preview jobs" which run after the main canvas is rendered could also store their results in the cache, and we could use these to render the previously out-of-frame portions of layers when zooming out or panning and composing the temporary images during render. 

So that's why I'm hesitant to introduce your changes right now.

vcloarec committed:
minimalist smooth mesh rendering when pan/zoom for mesh layer

vcloarec commented:
OK, let's give up (for 3.18) this temporal cache.
So now, an improvement only for pan/zoom.
The mReadyToCompose is set to true just after the scalar rendering and before other rendering to avoid increasing the temporal flickering, and for the same reason it is better to avoid the improvement for mesh frame and vector rendering ad that had been done for vector layer. 
Indeed, if we extend the time during no image is available, we will exceed the "map update interval" more often.
So better to not do more than this 3 lines before improvement of cache with temporal criteria.
